### PR TITLE
Add Cardano.Api.Tx generators

### PR DESCRIPTION
**Pull request by sevanspowell**
- [ ] Copied and extended Tx generators from the cardano-api library. We were unable to import them due to issues with cabal sublibs.
- [ ] Made a first pass at providing coverage tests for as many of the generators as possible.
    - Further note: I've decided to take a "best-effort" approach to providing coverage to generators. Coverage of generators is difficult, particularly for sum types and large product types. If you can see a better way of doing this, please let me know!

- Best reviewed commit-by-commit
    
### Issue Number

ADP-1147

### CODE REVIEW

**Comment by rvl** [comment]
@sevanspowell I have rebased this onto latest master.

**Comment by sevanspowell** [comment]
> @sevanspowell I have rebased this onto latest master.

Thank you!

**Comment by rvl** [comment]
How come there are two PRs for the same jira ticket with almost identical titles? #2975 and #2985.
I'm confused.

**Comment by sevanspowell** [comment]
> How come there are two PRs for the same jira ticket with almost identical titles? #2975 and #2985. I'm confused.

Sorry, that's my fault.

File: lib/core/test/unit/Cardano/Api/GenSpec.hs [hunk 1]
```diff
@@ -0,0 +1,1955 @@
...
+                    $ forAll (genTxInsCollateral AllegraEra)
+                    $ genTxInCollateralCoverage AllegraEra
+                it "MaryEra" $
+                    property
+                    $ forAll (genTxInsCollateral MaryEra)
+                    $ genTxInCollateralCoverage MaryEra
+                it "AlonzoEra" $
+                    property
+                    $ forAll (genTxInsCollateral AlonzoEra)
+                    $ genTxInCollateralCoverage AlonzoEra
```
Comment (rvl) [inline lib/core/test/unit/Cardano/Api/GenSpec.hs]: It should be possible to factor such code by using `AnyCardanoEra`. I put some code in the other PR - https://github.com/input-output-hk/cardano-wallet/pull/2975#discussion_r737222960.

**Review by rvl — APPROVED** [review]
Good thanks.
Please squash the branch though - it's actually harder to review the single commits for things like this.

**Comment by sevanspowell** [comment]
> it's actually harder to review the single commits for things like this.

Differing views on this. I will leave multiple commits for the review process and squash before bors. That way those that prefer multiple commits can review commit-by-commit, and those that don't can view the diff or squash locally.

File: lib/core/test/unit/Cardano/Api/GenSpec.hs (hunk 1 above)
Comment (sevanspowell) [reply lib/core/test/unit/Cardano/Api/GenSpec.hs]: Thank you. I tried to to this using existential quantification but then I had issues with the `era` quantifier escaping it's scope (skolem error). I'll try again with `AnyCardanoEra`.

File: lib/core/test/unit/Cardano/Api/GenSpec.hs [hunk 2]
```diff
@@ -1952,3 +1653,16 @@ signedCoverage name x = checkCoverage
     where
         veryLarge = fromIntegral (maxBound @Word32)
         verySmall = -fromIntegral (maxBound @Word32)
+
+forAllEras :: Applicative f => (AnyCardanoEra -> f a) -> f ()
+forAllEras f =
+    traverse_ f [ AnyCardanoEra ByronEra ]
```
Comment (rvl) [inline lib/core/test/unit/Cardano/Api/GenSpec.hs]:
```suggestion
    f (AnyCardanoEra ByronEra)
```

**Review by rvl — APPROVED** [review]

**Comment by sevanspowell** [comment]
bors cancel

**Pull request merged**